File: python/pathfinder/pyrvo/rvo_simulator.py
from typing import Tuple, List, Optional
from pathfinder.pyrvo.kd_tree import KDTree
from pathfinder.pyrvo.agent import Agent, linear_program2, linear_program3
from pathfinder.pyrvo.obstacle import Obstacle

class RVOSimulator:
    def __init__(self, neighbor_dist: float,
                 max_neighbors: int,
                 time_horizon: float,
                 time_horizon_obst: float,
                 radius: float,
                 max_speed: float):
        self._kd_tree = KDTree(self)
        self._default_agent = Agent(self, -1, None, (0.0, 0.0), (0.0, 0.0),
                                    neighbor_dist, max_neighbors, time_horizon, time_horizon_obst, radius, max_speed)
        # set default agent parameters
        # self._default_agent.set_parameters(max_neighbors, max_speed, neighbor_dist, radius, time_horizon, time_horizon_obst, (0.0, 0.0))
        self._agents: List[Agent] = []  # set agents empty array
        self._obstacles: List[Obstacle] = []  # also for obstacles
        self._force_update_agents = True

    def add_agent(self, position: Tuple[float, float],
                  radius: Optional[float] = None,
                  velocity: Tuple[float, float] = (0.0, 0.0),
                  neighbor_dist: Optional[float] = None,
                  max_neighbors: Optional[int] = None,
                  time_horizon: Optional[float] = None,
                  time_horizon_obst: Optional[float] = None,
                  max_speed: Optional[float] = None) -> int:
        agent = Agent(self, len(self._agents), self._default_agent, position, velocity, neighbor_dist, max_neighbors, time_horizon, time_horizon_obst, radius, max_speed)
        self._agents.append(agent)
        self._force_update_agents = True
        return len(self._agents) - 1

    def delete_agents(self, agent_indexes: List[int]):
        for i in range(len(agent_indexes)):
            agent_index = agent_indexes[len(agent_indexes) - 1 - i]
            self._agents.pop(agent_index)
        self._force_update_agents = True

    def get_agent(self, agent_index: int) -> Agent:
        return self._agents[agent_index]

    def get_all_agents(self) -> List[Agent]:
        return self._agents

    def add_obstacle(self, vertices: List[Tuple[float, float]]) -> int:
        if len(vertices) >= 2:
            obstacle_no: int = len(self._obstacles)
            for i in range(len(vertices)):
                obstacle = Obstacle()
                obstacle.set_point(vertices[i])
                if i != 0:
                    obstacle.set_prev_obstacle(self._obstacles[-1])
                    obstacle.get_prev_obstacle().set_next_obstacle(obstacle)

                if i == len(vertices) - 1:
                    obstacle.set_next_obstacle(self._obstacles[obstacle_no])
                    obstacle.get_next_obstacle().set_prev_obstacle(obstacle)

                obstacle.set_unit_dir(vertices[0 if (i == len(vertices) - 1) else i + 1], vertices[i])
                if len(vertices) == 2:
                    obstacle.set_is_convex_value(True)
                else:
                    obstacle.set_is_convex(vertices[(len(vertices) - 1 if i == 0 else i - 1)], vertices[i], vertices[0 if (i == len(vertices) - 1) else i + 1])

                obstacle.set_id(len(self._obstacles))
                self._obstacles.append(obstacle)
            return obstacle_no
                
        else:
            return -1

    def compute_agent_obstacles_neighbors(self, agent, range_square: float):
        self._kd_tree.compute_obstacle_neighbors(agent, range_square)

    def compute_agent_agents_neighbors(self, agent, range_square: float):
        self._kd_tree.compute_agent_neighbors(agent, range_square)

    def add_obstacle_object(self, obstacle: Obstacle):
        self._obstacles.append(obstacle)

    def do_step(self, delta_time: float, move_agents: bool = True):
        self._kd_tree.build_agent_tree(self._force_update_agents)
        self._force_update_agents = False
        for i in range(len(self._agents)):
            self._agents[i].compute_neighbors()
            self._agents[i].compute_new_velocity(delta_time)

        for i in range(len(self._agents)):
            self._agents[i].update(delta_time, move_agents)

    def get_agent_agent_neighbor(self, agent_index: int, neighbor_index: int) -> int:
        return self._agents[agent_index].get_agent_neighbors()[neighbor_index][1].get_id()

    def get_agent_max_neighbors(self, agent_index: int) -> int:
        return self._agents[agent_index].get_max_neighbors()

    def get_agent_max_speed(self, agent_index: int) -> float:
        return self._agents[agent_index].get_max_speed()

    def get_agent_neighbor_dist(self, agent_index: int) -> float:
        return self._agents[agent_index].get_neighbor_distance()

    def get_agent_num_agent_neighbors(self, agent_index: int) -> int:
        return len(self._agents[agent_index].get_agent_neighbors())

    def get_agent_num_obstacle_neighbors(self, agent_index) -> int:
        return len(self._agents[agent_index].get_obstacle_neighbors())

    def get_agent_obstacle_neighbor(self, agent_index: int, neighbor_index: int):
        return self._agents[agent_index].get_obstacle_neighbors()[neighbor_index][1].get_id()

    def get_agent_position(self, agent_index: int) -> Tuple[float, float]:
        return self._agents[agent_index].get_position()

    def get_agent_pref_velocity(self, agent_index: int) -> Optional[Tuple[float, float]]:
        return self._agents[agent_index].get_pref_velocity()

    def get_agent_radius(self, agent_index: int) -> float:
        return self._agents[agent_index].get_radius()

    def get_agent_time_horizon(self, agent_index: int) -> float:
        return self._agents[agent_index].get_time_horizon()

    def get_agent_time_horizon_obst(self, agent_index: int) -> float:
        return self._agents[agent_index].get_time_horizon_obst()

    def get_agent_velocity(self, agent_index: int) -> Tuple[float, float]:
        return self._agents[agent_index].get_velocity()

    def get_agents_count(self) -> int:
        return len(self._agents)

    def get_obstacles_count(self) -> int:
        return len(self._obstacles)

    def get_obstacle(self, obst_index: int) -> Obstacle:
        return self._obstacles[obst_index]

    def get_obstacle_vertex(self, vertex_index: int) -> Tuple[float, float]:
        return self._obstacles[vertex_index].get_point()

    def get_next_obstacle_vertex_index(self, vertex_index: int) -> int:
        return self._obstacles[vertex_index].get_next_obstacle().get_id()

    def get_prev_obstacle_vertex_index(self, vertex_index: int) -> int:
        return self._obstacles[vertex_index].get_prev_obstacle().get_id()

    def process_obstacles(self):
        self._kd_tree.build_obstacle_tree()

    def query_visibility(self, start: Tuple[float, float], end: Tuple[float, float], radius: float = 0.0) -> bool:
        return self._kd_tree.query_visibility(start, end, radius)

    def set_agent_defaults(self,
                           neighbor_dist: float,
                           max_neighbors: int,
                           time_horizon: float,
                           time_horizon_obst: float,
                           radius: float,
                           max_speed: float):
        self._default_agent.set_parameters(max_neighbors, max_speed, neighbor_dist, radius, time_horizon, time_horizon_obst, (0.0, 0.0))

    def set_agent_max_neighbors(self, agent_index: int, max_neighbors: int):
        self._agents[agent_index].set_max_neighbors(max_neighbors)

    def set_agent_max_speed(self, agent_index: int, max_speed: float):
        self._agents[agent_index].set_max_speed(max_speed)

    def set_agent_neighbor_dist(self, agent_index: int, neighbor_dist: float):
        self._agents[agent_index].set_neighbor_distance(neighbor_dist)

    def set_agent_position(self, agent_index: int, position: Tuple[float, float]):
        self._agents[agent_index].set_position(position)

    def set_agent_pref_velocity(self, agent_index: int, velocity: Tuple[float, float]):
        self._agents[agent_index].set_pref_velocity(velocity)

    def set_agent_radius(self, agent_index: int, radius: float):
        self._agents[agent_index].set_radius(radius)

    def set_agent_time_horizon(self, agent_index: int, time_horizon: float):
        self._agents[agent_index].set_time_horizon(time_horizon)

    def set_agent_time_horizon_obst(self, agent_index: int, time_horizon_obst: float):
        self._agents[agent_index].set_time_horizon_obst(time_horizon_obst)

    def set_agent_velocity(self, agent_index: int, velocity: Tuple[float, float]):
        self._agents[agent_index].set_velocity(velocity)

File: assemblyscript/assembly/rvo/rvo_simulator.ts
import { KDTree } from "./rvo_kd_tree";
import { Obstacle } from "./rvo_obstacle";
import { Agent } from "./rvo_agent";
import { Vector2 } from "../common/vector2";
import { List } from "../common/list";
import { log_message, is_in_array } from "../common/utilities";

export class RVOSimulator{
    private m_agents: List<Agent>;
    private m_agents_count: i32;
    private m_kd_tree: KDTree;
    private m_obstacles: List<Obstacle>;

    private m_neighbor_dist: f32;
    private m_max_neighbors: i32;
    private m_time_horizon: f32;
    private m_time_horizon_obst: f32;
    private m_radius: f32;
    private m_max_speed: f32;

    private m_force_update_agents: bool;

    constructor(neighbor_dist: f32,
                max_neighbors: i32,
                time_horizon: f32,
                time_horizon_obst: f32,
                radius: f32,
                max_speed: f32) {
        this.m_neighbor_dist = neighbor_dist;
        this.m_max_neighbors = max_neighbors;
        this.m_time_horizon = time_horizon;
        this.m_time_horizon_obst = time_horizon_obst;
        this.m_radius = radius;
        this.m_max_speed = max_speed;
        this.m_kd_tree = new KDTree();
        this.m_agents = new List<Agent>();
        this.m_agents_count = 0;
        this.m_obstacles = new List<Obstacle>();

        this.m_force_update_agents = true;
        this.m_kd_tree.set_simulator(this);
    }

    @inline
    add_agent(position_x: f32, position_y: f32): i32{
        return this.add_agent_ext(position_x, position_y, 
                                  0.0, 0.0,
                                  this.m_radius,
                                  this.m_neighbor_dist,
                                  this.m_max_neighbors,
                                  this.m_time_horizon,
                                  this.m_time_horizon_obst,
                                  this.m_max_speed);
    }

    @inline
    add_agent_ext(position_x: f32, position_y: f32,
                  velocity_x: f32, velocity_y: f32,
                  radius: f32,
                  neighbor_dist: f32,
                  max_neighbors: i32,
                  time_horizon: f32,
                  time_horizon_obst: f32,
                  max_speed: f32): i32{
        let agent = new Agent(this, this.m_agents.length,
                              new Vector2(position_x, position_y),
                              new Vector2(velocity_x, velocity_y),
                              neighbor_dist,
                              max_neighbors,
                              time_horizon,
                              time_horizon_obst,
                              radius,
                              max_speed);
        this.m_agents.push(agent);
        this.m_agents_count++;
        this.m_force_update_agents = true;
        return this.m_agents.length - 1;
    }

    @inline
    delete_agent(agent_index: i32): void{
        let to_delete = new List<i32>();
        to_delete.push(agent_index);
        this.delete_agents(to_delete);
    }

    @inline
    delete_agents(agent_indexes: List<i32>): void{
        var old_index: i32 = 0;
        const agents_count = this.m_agents.length;
        var new_agents = new List<Agent>(agents_count - agent_indexes.length);
        while(old_index < agents_count){
            if(!is_in_array(old_index, agent_indexes)){
                new_agents.push(this.m_agents[old_index]);
            }
            old_index++;
        }
        this.m_agents = new_agents;
        this.m_agents_count = this.m_agents.length;
        this.m_force_update_agents = true;
    }

    @inline
    get_agent(index: i32): Agent{
        return this.m_agents[index];
    }

    @inline
    get_agents_count(): i32{
        return this.m_agents_count;
    }

    //the same as add_obstacle, but input is plain float array
    add_obstacle_array(vertices: StaticArray<f32>): i32{
        let vec2_vertices = new StaticArray<Vector2>(vertices.length / 2);
        for(let  i = 0, len = vec2_vertices.length; i < len; i++){
            unchecked(vec2_vertices[i]).set_values(unchecked(vertices[2*i]), unchecked(vertices[2*i + 1]));
        }
        return this.add_obstacle(vec2_vertices);
    }

    add_obstacle(vertices: StaticArray<Vector2>): i32{
        const vertices_length = vertices.length;
        if (vertices_length < 2) {
            return -1;
        }

        const obstacle_no: i32 = this.m_obstacles.length;

        for (let i = 0; i < vertices_length; i++) {
            var obstacle = new Obstacle();
            obstacle.set_point(unchecked(vertices[i]));

            if (i != 0) {
                obstacle.set_prev_obstacle(this.m_obstacles[this.m_obstacles.length - 1]);
                let po = obstacle.get_prev_obstacle();
                if(po){
                    po.set_next_obstacle(obstacle);
                }
            }

            if (i == vertices_length - 1) {
                obstacle.set_next_obstacle(this.m_obstacles[obstacle_no]);
                let no = obstacle.get_next_obstacle();
                if(no){
                    no.set_prev_obstacle(obstacle);
                }
            }

            const p1 = unchecked(vertices[(i == vertices_length - 1 ? 0 : i + 1)]);
            const p2 = unchecked(vertices[i]);
            obstacle.set_unit_dir_point(p1.x(), p1.y(), p2.x(), p2.y());

            if (vertices_length == 2) {
                obstacle.set_is_convex_value(true);
            }
            else {
                const p3 = unchecked(vertices[(i == 0 ? vertices_length - 1 : i - 1)]);
                obstacle.set_is_convex(p3.x(), p3.y(), p2.x(), p2.y(), p1.x(), p1.y());
            }
            obstacle.set_id(this.m_obstacles.length);
            this.m_obstacles.push(obstacle);
        }

        return obstacle_no;
    }

    @inline
    get_obstacles_count(): i32{
        return this.m_obstacles.length;
    }

    @inline
    get_obstacle(index: i32): Obstacle{
        return this.m_obstacles[index];
    }

    @inline
    compute_agent_obstacles_neighbors(agent: Agent, range_square: f32): void{
        this.m_kd_tree.compute_obstacle_neighbors(agent, range_square);
    }

    @inline
    compute_agent_agents_neighbors(agent: Agent, range_square: f32): void{
        this.m_kd_tree.compute_agent_neighbors(agent, range_square);
    }

    @inline
    add_obstacle_object(obstacle: Obstacle): void{
        this.m_obstacles.push(obstacle);
    }

    do_step_default(delta_time: f32): void{
        this.do_step(delta_time, true);
    }    

    do_step(delta_time: f32, move_agents: bool = true): void{
        this.m_kd_tree.build_agent_tree(this.m_force_update_agents);
        this.m_force_update_agents = false;
        for(let i = 0, len = this.m_agents_count; i < len; i++){
            let agent = this.m_agents[i];
            agent.compute_neighbors();
            agent.compute_new_velocity(delta_time);
        }

        for(let i = 0, len = this.m_agents_count; i < len; i++){
            this.m_agents[i].update(delta_time, move_agents);
        }
    }

    @inline
    get_agent_agent_neighbor(agent_index: i32, neighbor_index: i32): i32{
        return this.m_agents[agent_index].get_agent_agent_neighbor(neighbor_index).get_id();
    }

    @inline
    get_agent_max_neighbors(agent_index: i32): i32{
        return this.m_agents[agent_index].get_max_neighbors();
    }

    @inline
    get_agent_max_speed(agent_index: i32): f32{
        return this.m_agents[agent_index].get_max_speed();
    }

    @inline
    get_agent_neighbor_dist(agent_index: i32): f32{
        return this.m_agents[agent_index].get_neighbor_distance();
    }

    @inline
    get_agent_num_agent_neighbors(agent_index: i32): i32{
        return this.m_agents[agent_index].get_agent_neighbors_count();
    }

    @inline
    get_agent_num_obstacle_neighbors(agent_index: i32): i32{
        return this.m_agents[agent_index].get_num_obstacle_neighbors();
    }

    @inline
    get_agent_obstacle_neighbor(agent_index: i32, neighbor_index: i32): i32{
        return this.m_agents[agent_index].get_obstacle_neighbor(neighbor_index).get_id();
    }

    @inline
    get_agent_position(agent_index: i32): Vector2{
        return this.m_agents[agent_index].get_position();
    }

    //return one plain float array with 2d-positions of all agents in the simulator
    @inline
    get_agents_positions(): StaticArray<f32> {
        let to_return = new StaticArray<f32>(2 * this.m_agents_count);
        for(let i = 0, len = this.m_agents_count; i < len; i++){
            let p = this.get_agent_position(i);
            unchecked(to_return[2*i] = p.x());
            unchecked(to_return[2*i + 1] = p.y());
        }
        return to_return;
    }

    @inline
    get_agent_pref_velocity(agent_index: i32): Vector2{
        return this.m_agents[agent_index].get_pref_velocity();
    }

    @inline
    get_agent_radius(agent_index: i32): f32{
        return this.m_agents[agent_index].get_radius();
    }

    @inline
    get_agent_time_horizon(agent_index: i32): f32{
        return this.m_agents[agent_index].get_time_horizon();
    }

    @inline
    get_agent_time_horizon_obst(agent_index: i32): f32{
        return this.m_agents[agent_index].get_time_horizon_obst();
    }

    @inline
    get_agent_velocity(agent_index: i32): Vector2{
        return this.m_agents[agent_index].get_velocity();
    }

    //return one plain float array with velocities of all agents
    @inline
    get_agents_velocities(): StaticArray<f32> {
        let to_return = new StaticArray<f32>(2 * this.m_agents_count);
        for(let i = 0, len = this.m_agents_count; i < len; i++){
            let v = this.get_agent_velocity(i);
            unchecked(to_return[2*i] = v.x());
            unchecked(to_return[2*i + 1] = v.y());
        }
        return to_return;
    }

    @inline
    get_obstacle_vertex(vertex_index: i32): Vector2{
        return this.m_obstacles[vertex_index].get_point();
    }

    @inline
    get_next_obstacle_vertex_index(vertex_index: i32): i32{
        let obst = this.m_obstacles[vertex_index].get_next_obstacle();
        if(obst){
            return obst.get_id();
        }
        else{
            return -1;
        }
    }

    @inline
    get_prev_obstacle_vertex_index(vertex_index: i32): i32{
        let obst = this.m_obstacles[vertex_index].get_prev_obstacle();
        if(obst){
            return obst.get_id();
        }
        else{
            return -1;
        }
    }

    @inline
    process_obstacles(): void{
        this.m_kd_tree.build_obstacle_tree();
    }

    @inline
    query_visibility(start_x: f32, start_y: f32,
                     end_x: f32, end_y: f32,
                     radius: f32): bool{
        return this.m_kd_tree.query_visibility(new Vector2(start_x, start_y), new Vector2(end_x, end_y), radius);
    }

    @inline
    set_agent_max_neighbors(agent_index: i32, max_neighbors: i32): void{
        this.m_agents[agent_index].set_max_neighbors(max_neighbors);
    }

    @inline
    set_agent_max_speed(agent_index: i32, max_speed: f32): void{
        this.m_agents[agent_index].set_max_speed(max_speed);
    }

    @inline
    set_agent_neighbor_dist(agent_index: i32, neighbor_dist: f32): void{
        this.m_agents[agent_index].set_neighbor_distance(neighbor_dist);
    }

    @inline
    set_agent_position(agent_index: i32, position_x: f32, position_y: f32): void{
        this.m_agents[agent_index].set_position(new Vector2(position_x, position_y));
    }

    //input is array with 2d-positions of all agents
    @inline
    set_agents_positions(positions: StaticArray<f32>): void{
        const input_positions = positions.length / 2;
        const count = input_positions < this.m_agents_count ? input_positions : this.m_agents_count;
        for(let i = 0; i < count; i++){
            this.set_agent_position(i, unchecked(positions[2*i]), unchecked(positions[2*i + 1]));
        }
    }

    @inline
    set_agent_pref_velocity(agent_index: i32, velocity_x: f32, velocity_y: f32): void{
        this.m_agents[agent_index].set_pref_velocity(new Vector2(velocity_x, velocity_y));
    }

    //input is array with prefered 2d-velocities for all agents
    @inline
    set_agents_pref_velocities(velocities: StaticArray<f32>): void{
        const input_velocities = velocities.length / 2;
        const count = input_velocities < this.m_agents_count ? input_velocities : this.m_agents_count;
        for(let i = 0; i < count; i++){
            this.set_agent_pref_velocity(i, unchecked(velocities[2*i]), unchecked(velocities[2*i + 1]));
        }
        //for other agents set zero
        for(let i = 0; i < this.m_agents_count - count; i++){
            this.set_agent_pref_velocity(count + i, 0.0, 0.0);
        }
    }

    @inline
    set_agent_radius(agent_index: i32, radius: f32): void{
        this.m_agents[agent_index].set_radius(radius);
    }

    @inline
    set_agent_time_horizon(agent_index: i32, time_horizon: f32): void{
        this.m_agents[agent_index].set_time_horizon(time_horizon);
    }

    @inline
    set_agent_time_horizon_obst(agent_index: i32, time_horizon_obst: f32): void{
        this.m_agents[agent_index].set_time_horizon_obst(time_horizon_obst);
    }

    @inline
    set_agent_velocity(agent_index: i32, velocity_x: f32, velocity_y: f32): void{
        this.m_agents[agent_index].set_velocity(new Vector2(velocity_x, velocity_y));
    }
}
